### ISSUE TITLE
The slow running script on ShARC for which 96 hours is not enough

### ISSUE THREAD

**Issue by mikecroucher** [ooominds/RSE-ticketing#2]
I assume that the script in question is in the shared area on ShARC?

What is the full path to the script that did not complete on a 32 core node within the 96 hour time limit?

**Comment by kuchenrolle** [comment]
The full path to the directory is /shared/ooominds1/Shared/noun_allomorphy/ and the script is scripts/train.py, submitted with scripts/train.sh.

You mentioned that my process was not making use of the allocated cores. I tested this and indeed the CPU time on a test file increased with more cores, due to the overhead. I've run the same python script with the same input on another server and it immediately used the nuber of cores it was supposed to, so I take it that the problem is either with the bash script or somehow the pyndl is prohibited from making use of the cores on sharc. The code for pyndl can be found here: https://github.com/quantling/pyndl (installing via pip didn't work for me, so I cloned the repository and installed with python setup.py install --user)

I created a git repository for figuring out what the problem is and benchmarking pyndl, it lives here: https://github.com/ooominds/benchmark_pyndl/

**Comment by trallard** [comment]
In an attempt to gather as much information as possible...  @kuchenrolle could you fill in the missing details here, please?

## Steps to Reproduce


1.
2.
3.
4.

## Your Environment

* Environment name and version (e.g. Python 3.5, R 4.3):
* Additional libraries you are using: [pyndl](https://github.com/quantling/pyndl)
* Link to your project:

## Extras:
* Benchmarking repository: [https://github.com/ooominds/benchmark_pyndl/](https://github.com/ooominds/benchmark_pyndl/)

- [ ] Is the benchmark still a work in progress? Tick is yes

**Comment by kuchenrolle** [comment]
## Steps to Reproduce
1. qsub /mnt/shared/ooominds1/noun_allomorphy/scripts/train.sh (change email adress)

## Your Environment
* Environment name and version (e.g. Python 3.5, R 4.3): Anaconda3 (4.2.0), iPython 6.1.0, Python 3.6.2
* Any additional libraries or dependencies you are using: [pyndl](https://github.com/quantling/pyndl)
* Link to your project: [noun_allomorphy](https://github.com/ooominds/noun_allomorphy/)

## Extras:
* Benchmarking repository: https://github.com/ooominds/benchmark_pyndl/
- [x] Is the benchmark still a work in progress? Tick is yes

**Comment by mikecroucher** [comment]
The script activates an Anaconda module called sharc 

```
source activate sharc
```

Presumably this is one that you've created yourself?  Do you have the details of how you created it?  I can't run the script otherwise.

Useful information can be obtained by doing

```
source activate sharc
conda list
```

@kuchenrolle Would you mind doing this please and paste the output here

**Comment by mikecroucher** [comment]
It is not obvious what the input and output is since there are no comments in either the bash script or the python script 

```
ipython3 train.py "/shared/ooominds1/Shared/noun_allomorphy/data/event_files/tri
.gz" "/shared/ooominds1/Shared/noun_allomorphy/data/matrices/tri.nc"
```

This makes it difficult to test since I'd be worried I might be overwriting your output in the course of my testing.  I guess the output is `tri.nc` -- can you confirm please?

**Comment by kuchenrolle** [comment]
Yes, source activate loads my conda environment called sharc and the output of the script is the tri.nc file, while the event file (tri.gz) will not be changed. I added comments to train.py. Here's the list of installed packages:

```
alabaster                 0.7.10                    <pip>
Babel                     2.5.3                     <pip>
certifi                   2016.2.28                py36_0  
cycler                    0.10.0                    <pip>
cymem                     1.31.2                   py36_0  
Cython                    0.27.3                    <pip>
cytoolz                   0.8.2                    py36_0  
de-core-news-md           1.0.0                     <pip>
decorator                 4.1.2                    py36_0  
dill                      0.2.6                    py36_0  
docopt                    0.6.2                     <pip>
docutils                  0.14                      <pip>
en-core-web-lg            2.0.0                     <pip>
en-core-web-sm            1.2.0                     <pip>
ftfy                      4.4.3                    py36_0  
html5lib                  0.9999999                py36_0  
imagesize                 0.7.1                     <pip>
ipython                   6.1.0                    py36_0  
ipython_genutils          0.2.0                    py36_0  
jedi                      0.10.2                   py36_2  
Jinja2                    2.10                      <pip>
libgcc                    5.2.0                         0  
MarkupSafe                1.0                       <pip>
matplotlib                2.1.1                     <pip>
mkl                       2017.0.3                      0  
msgpack-numpy             0.4.1                      py_0    conda-forge
msgpack-python            0.4.8                    py36_0  
murmurhash                0.26.4                   py36_0  
netCDF4                   1.3.1                     <pip>
numpy                     1.13.1                   py36_0  
numpydoc                  0.7.0                     <pip>
openssl                   1.0.2l                        0  
pandas                    0.20.3                   py36_0  
path.py                   10.3.1                   py36_0  
pathlib                   1.0.1                    py36_0  
pexpect                   4.2.1                    py36_0  
pickleshare               0.7.4                    py36_0  
pip                       9.0.1                    py36_1  
plac                      0.9.6                    py36_0  
preshed                   1.0.0                    py36_0  
prompt_toolkit            1.0.15                   py36_0  
ptyprocess                0.5.2                    py36_0  
pygments                  2.2.0                    py36_0  
pyparsing                 2.2.0                     <pip>
python                    3.6.2                         0  
python-dateutil           2.6.1                    py36_0  
pytz                      2017.2                   py36_0  
readline                  6.2                           2  
regex                     2017.4.5                 py36_0  
requests                  2.14.2                   py36_0  
setuptools                36.4.0                   py36_1  
simplegeneric             0.8.1                    py36_1  
six                       1.10.0                   py36_0  
snowballstemmer           1.2.1                     <pip>
spacy                     1.8.2                    py36_0  
Sphinx                    1.6.6                     <pip>
sphinxcontrib-websupport  1.0.1                     <pip>
sqlite                    3.13.0                        0  
termcolor                 1.1.0                    py36_0  
thinc                     6.5.2                    py36_0  
thinc                     6.10.2                    <pip>
tk                        8.5.18                        0  
toolz                     0.8.2                    py36_0  
tqdm                      4.15.0                   py36_0  
traitlets                 4.3.2                    py36_0  
ujson                     1.35                     py36_0  
wcwidth                   0.1.7                    py36_0  
wheel                     0.29.0                   py36_0  
wrapt                     1.10.11                  py36_0  
xarray                    0.10.0                    <pip>
xz                        5.2.3                         0  
zlib                      1.2.11                        0  
```

To reproduce, it should be sufficient to install pyndl and its dependencies and change the bash scripts to load your environment:

```
git clone https://github.com/quantling/pyndl.git
cd pyndl
python setup.py install --user
```

On Windows it might be necessary to first install the dependencies:
```
source activate <yourCondaEnv>
conda install numpy cython pandas xarray netCDF4 numpydoc pip
```

**Comment by mikecroucher** [comment]
I created the environment as follows 

```
conda create -n ooominds_test Python=3.6
source activate ooominds_test
conda install ipython
conda install pandas
conda install netCDF4
conda install Cython
conda install xarray
```

This bug here https://github.com/quantling/pyndl/issues/106 meant that I couldn't install pyndl from source so, like you, I installed via github.

I have your script working and I can confirm that it is spawning as many Ipython processes as we expect.  That is, it IS working in parallel.  To determine this, I ran the `ipython3` command interactively in the background (appended `&` to the end of the command) and ran `top`.

So, now we need to dig further to see what is going on

We can turn on verbose output in pyndl which we should do so we can see how far along it is getting in the computation.

```
weights = ndl.ndl(events=_events,
                      alpha=0.1, betas=(0.1, 0.1),
                      method="openmp", number_of_threads=32,
                      remove_duplicates=True,verbose=True)
```

What they decide to print isn't always useful but it's better than a blank screen.
The first time consuming thing it does is run a function called `_job_cues_outcomes()` in count.py: https://github.com/quantling/pyndl/blob/84587aa5d70ce40be7473af2a590b77d55522155/pyndl/count.py

The verbose output wasn't very useful in that all it did was print a bunch of .....
Will I be seeing dots for a few minutes or until the heat death of the Universe?

I hacked up a progress monitor for this function and determined that the variable `nn` gets as high as **60.5 million**. This is probably proportional to your problem size.

Have you ever run pyndl with this size of problem before?

Would it be possible to provide a smaller input file please?

**Comment by kuchenrolle** [comment]
Yes, nn is the number of events (which should equal the number of lines in the input binary tri.gz file). You can check that in python with
```
import gzip
num_events = 0
with gzip.open("/shared/ooominds1/Shared/noun_allomorphy/data/event_files/tri.gz","rt") as events:
    for line in events: num_events+=1
num_events
```
or without python using unix' zgrep counting linebreaks:
`zgrep -Ec "$" /shared/ooominds1/Shared/noun_allomorphy/data/event_files/tri.gz`

What count.cues_outcomes does is run through the whole event file, counting the number of events and all the rows (cues) and columns (outcomes) of the weight matrix that it will produce (in parallel, calling _job_cues_outcomes and using python's multiprocessing - i.e. this is not the openmp part that actually does the matrix calculations). With verbose=True it will print a dot for every 100k lines processed.

I have run pyndl (though earlier versions) with similarly sized inputs and substantially bigger ones.

You can downsize the input file using this (maybe 2 million would be a good size for testing):
```
import gzip
with gzip.open("/shared/ooominds1/Shared/noun_allomorphy/data/event_files/tri.gz","rt") as events:
    gzip.open("/shared/ooominds1/Shared/noun_allomorphy/data/event_files/tri_test.gz","wt") as events_new:
        for i in range(num_events):
            line = events.readline()
            events_new.write(line)
```

PS: Could you install htop on sharc? That would be very helpful for monitoring resource utilization as it enhances the top interface.

**Comment by mikecroucher** [comment]
I've discovered the first problem.  It seems that the code uses `/tmp` a lot and it gets full.  The code responds by freezing.  

If we had left it running, it would have run forever with no results.

The fix is to add the following to the bash submission script before running the `ipython3` command.

```
export TMPDIR=/scratch
```

This gets things a lot further but the job is now being killed by the scheduler a couple of hours in. `qacct` output from the most recent attempt is 

```
[fe1mpc@sharc-login2 fe1mpc]$ qacct -j 874193
==============================================================
qname        rse.q               
hostname     sharc-node142.shef.ac.uk
group        fe                  
owner        fe1mpc              
project      rse                 
department   defaultdepartment   
jobname      trainNDL            
jobnumber    874193              
taskid       undefined
account      sge                 
priority     0                   
qsub_time    Wed Jan 31 03:20:41 2018
start_time   Wed Jan 31 03:20:56 2018
end_time     Wed Jan 31 05:57:07 2018
granted_pe   smp                 
slots        32                  
failed       37  : qmaster enforced h_rt, h_cpu, or h_vmem limit
exit_status  137                  (Killed)
ru_wallclock 9371s
ru_utime     2.470s
ru_stime     0.998s
ru_maxrss    29.586KB
ru_ixrss     0.000B
ru_ismrss    0.000B
ru_idrss     0.000B
ru_isrss     0.000B
ru_minflt    55319               
ru_majflt    0                   
ru_nswap     0                   
ru_inblock   0                   
ru_oublock   32                  
ru_msgsnd    0                   
ru_msgrcv    0                   
ru_nsignals  0                   
ru_nvcsw     4356                
ru_nivcsw    39                  
cpu          261742.740s
mem          854.088TBs
io           4.531TB
iow          0.000s
maxvmem      160.003GB
arid         undefined
ar_sub_time  undefined
```

I suspect a lack of memory so I am running again on on of our large memory nodes.

**Comment by kuchenrolle** [comment]
The problem with /tmp comes up when ndl splits the event file into a bunch of integer-mapped sublists [here](https://github.com/quantling/pyndl/blob/master/pyndl/ndl.py#L158). Adding an optional temporary_directory argument to the ndl function and passing it over to TemporaryDirectory seems like a good solution, I'll put in a pull request on pyndl later.

Does "things get a lot further" mean it gets to the point where the calculations are actually being done and do they run in parallel?

The memory problem actually occurred on the tuebingen servers as well and the script, when run on the full file, required 400G at some point (at which point I killed it). This apparently happened only once before and they couldn't figure out what was happening back then, so my first guess was that it's a problem with the event file, in that maybe something happens on a special character (because the file contains polish diacritics, dots, dashes and numbers), but I haven't found anything on that front.

**Comment by mikecroucher** [comment]
Well the memory problem really is a memory problem! 
480Gig isn't enough.

```
Job 874353 (trainNDL) Aborted
Exit Status      = 137
Signal           = KILL
User             = fe1mpc
Queue            = rse.q@sharc-node123.shef.ac.uk
Host             = sharc-node123.shef.ac.uk
Start Time       = 01/31/2018 09:29:44
End Time         = 01/31/2018 15:08:39
CPU              = 80:59:57
Max vmem         = 480.008G
failed qmaster enforced h_rt, h_cpu, or h_vmem limit because:
job 874353.1 died through signal KILL (9)
```

I'll make one last try with 755 Gig which is the maximum we can support.

**Comment by kuchenrolle** [comment]
Just to be clear: this is not the usual or expected pyndl behaviour. Memory use should be a small fraction of this, there's something going terribly wrong somewhere!

**Comment by mikecroucher** [comment]
In verbose mode, the program submits a bunch of jobs...127 in total

```
....
submitted job 126
submitted job 127
```

and then slowly but surely we get `job finished` 127 times.  
It's after the last `job finished` that the scheduler kills the job due to memory usage.  

I'm still waiting on the 755Gig job.  I know it shouldn't need that much but curiosity has got the better of me.

**Comment by kuchenrolle** [comment]
Can't say that I'm not curious too and wouldn't have done the same thing. (;

Are the jobs all submitted and running at once? Or are some of them for counting (32), for creating the binary files (32) and then for learning (32) with some additional ones that manage these? Because ndl.ndl calls count.cues_outcomes, then preprocess.create_binary_event_files and finally ndl_parallel.learn_inplace with the number of threads specified, so I don't think it should ever run that many jobs in parallel. If that is the case, though, you could check whether switching from openmp to pure python threading makes any difference (change method="openmp" to method="threading" in the ndl call).

**Comment by mikecroucher** [comment]
No idea what it's doing at the moment.

I just took your script and ran it with verbose set to `True`

755 Gigabytes wasn't enough.  So we definitely need to figure out what's going on here.

**Comment by derNarr** [comment]
I start debugging a similar problem. The event file is 127 MB. The binary files are around 867 MB. The weight matrix needs roughly 127 MB. The problem has 33,347,683 events, 24,663 unique outcomes, and 5,407 unique cues. The machine has 40 physical cores and 1TB of memory.

Counting the events, cues and outcomes and creating the binary files works fine. (It is done within 20 minutes.)

After starting the learning the memory footprint increases drastically and not all cores that should be used, are used.

The behaviour is the same regardless of using the `threading` or the `openmp` method.

If I get some new insights, I let you know.

**Comment by mikecroucher** [comment]
@kuchenrolle Can we get a subset of the input data for which this code completes?

**Comment by kuchenrolle** [comment]
I put an event file here: /shared/ooominds1/Shared/tmp/test.gz

pyndl runs fine for that, it's significantly smaller though.

**Comment by derNarr** [comment]
I start debugging a similar problem. The event file is 127 MB. The binary files are around 867 MB. The weight matrix needs roughly 127 MB. The problem has 33,347,683 events, 24,663 unique outcomes, and 5,407 unique cues. The machine has 40 physical cores and 1TB of memory.

Counting the events, cues and outcomes and creating the binary files works fine. (It is done within 20 minutes.)

After starting the learning the memory footprint increases drastically and not all cores that should be used, are used.

The behaviour is the same regardless of using the `threading` or the `openmp` method.

If I get some new insights, I let you know.

**Comment by derNarr** [comment]
I might have found the bug. I will let you know asap.

**Comment by derNarr** [comment]
Ok, the memory leak should be fixed now. There was one `malloc` to much in the `ndl_parallel.pyx` file. When I spotted it, I could not believe that i have not seen it before :)

Line 157 in `ndl_parallel.pyx` in the following diff https://github.com/quantling/pyndl/commit/38e76b72cd1ea73fb6a38766ed4ff3f4327f88cd#diff-ec65aa81c6082f12ad722b470ef15009L157

The new pyndl version is merged into the master on github and uploaded onto pypi.

**Comment by derNarr** [comment]
Let me know, if this fixes your problem. Thanks!

**Comment by kuchenrolle** [comment]
Awesome work!

I just updated pyndl (pip install still fails) and started a job, will report back once it's run through!

**Comment by trallard** [comment]
@kuchenrolle  any idea if this solves the problem?

**Comment by kuchenrolle** [comment]
Went through, looks good. This solved the problem and this can be closed.

Thanks, Tino!
